guava/src/com/google/common/cache/CacheBuilder.java

```diff
@@ -650,6 +651,35 @@ Strength getValueStrength() {
     return this;
   }
 
+  /**
+   * Specifies that each entry should be automatically removed from the cache once a fixed duration
+   * has elapsed after the entry's creation, or the most recent replacement of its value.
+   *
+   * <p>When {@code duration} is zero, this method hands off to {@link #maximumSize(long)
+   * maximumSize}{@code (0)}, ignoring any otherwise-specified maximum size or weight. This can be
+   * useful in testing, or to disable caching temporarily without a code change.
+   *
+   * <p>Expired entries may be counted in {@link Cache#size}, but will never be visible to read or
+   * write operations. Expired entries are cleaned up as part of the routine maintenance described
+   * in the class javadoc.
+   *
+   * @param duration the length of time after an entry is created that it should be automatically
+   *                 removed
+   * @return this {@code CacheBuilder} instance (for chaining)
+   * @throws IllegalArgumentException if {@code duration} is negative
+   * @throws IllegalStateException is the time to live or time to idle was already set
+   */
+  public CacheBuilder<K, V> expireAfterWrite(Duration duration) {
+    checkState(
```

Comment:
Unless we're going to plumb the Duration instance all the way through the cache internals (which would require a serious re-write), then I think we should just call `expireAfterWrite(duration.toNanos(), TimeUnit.NANOSECONDS)` here.

Ditto for the other overloads.

Reply:
Your right, I didn't see that at first. That makes way more sense. I will make the modification soon.

Reply:
I'm about to submit a change to add these overloads, as well as updates for all of the tests (so this pull request can probably be closed).
Thanks for the push in the right direction!

Reply:
I noticed that we don't check for overflow after ns conversion. I'm planning on fixing that sometime soon in Caffeine, just in case. You might want to bundle that into your CL.

Reply:
I'm not quite sure I follow.

I was planning on just letting `duration.toNanos()` blow up if passed > ~292 years (and document as such)

Reply:
sorry, I meant for the `TimeUnit` one being delegated to. I thought one could write `expireAfterWrite(Long.MAX_VALUE, TimeUnit.DAYS)` without it blowing up due to the negative check preceding the conversion. But it looks like the conversion doesn't overflow and saturates instead.
```
java> TimeUnit.DAYS.toNanos(Long.MAX_VALUE)
java.lang.Long res1 = 9223372036854775807
```

Reply:
Yep, the saturation behavior of `TimeUnit.toFoo()` is extremely surprising...I'm shocked it hasn't caused more bugs actually :-)